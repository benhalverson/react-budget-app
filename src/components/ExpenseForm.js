import React, { useState } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

function useInput({ initialValue, name }) {
  const [value, setValue] = useState(initialValue);
  function onChange(e) {
    const updated = e.target.value;
    setValue(updated);
  }
  return [value, onChange];
}

export default function ExpenseForm({ expense, onSubmit }) {
  const [description, onDescriptionChange] = useInput(
    expense ? expense.description : ''
  );
  const [note, onNoteChange] = useInput(expense ? expense.note : '');
  const [category, onCategoryChange] = useInput(
    expense ? expense.category : ''
  );
  const [amount, setAmount] = useState(
    expense ? (expense.amount / 100).toString() : ''
  );
  const [createdAt, setCreatedAt] = useState(
    expense ? moment(expense.createdAt) : moment()
  );
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState('');

  function onAmountChange(e) {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  }
  function onDateChange(createdAt) {
    if (createdAt) {
      setCreatedAt(createdAt);
    }
  }
  function onFocusChange({ focused }) {
    setCalendarFocused(focused);
  }
  function onFormSubmit(e) {
    e.preventDefault();

    if (!description || !amount) {
      setError('Please provide description, category, and amount.');
    } else {
      setError('');
      onSubmit({
        description,
        amount: parseFloat(amount) * 100,
        category,
        createdAt: createdAt.valueOf(),
        note,
      });
    }
    console.log(category);
  }

  return (
    <form className='form' onSubmit={onFormSubmit}>
      {error && <p className='form__error'>{error}</p>}
      <input
        type='text'
        placeholder='Description'
        autoFocus
        className='text-input'
        value={description}
        onChange={onDescriptionChange}
      />
      <input
        type='text'
        placeholder='Category'
        autoFocus
        className='text-input'
        value={category}
        onChange={onCategoryChange}
      />
      <input
        type='text'
        placeholder='Amount'
        className='text-input'
        value={amount}
        onChange={onAmountChange}
      />
      <SingleDatePicker
        date={createdAt}
        onDateChange={onDateChange}
        focused={calendarFocused}
        onFocusChange={onFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
      <textarea
        placeholder='Add a note for your expense (optional)'
        className='textarea'
        value={note}
        onChange={onNoteChange}
      />
      <div>
        <button className='button'>Save Expense</button>
      </div>
    </form>
  );
}
