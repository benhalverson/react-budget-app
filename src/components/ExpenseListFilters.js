import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          startDateId={'START_DATE'}
          endDateId={'END_DATE'}
        />
      </div>
    );
  }
}
const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(mapStateToProps, {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
})(ExpenseListFilters);
