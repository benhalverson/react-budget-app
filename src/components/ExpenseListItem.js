import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ expense: { description, id, amount, createdAt }}) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {numeral(amount / 100).format('$0,0.00')} -{' '}
      {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;
