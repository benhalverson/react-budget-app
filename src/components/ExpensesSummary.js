import React from 'react';
import { useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

const ExpensesSummary = () => {
  const { expenseCount, expensesTotal } = useSelector(({ expenses, filters }) => {
    const visibleExpenses = selectExpenses(expenses, filters);
    return {
      expenseCount: visibleExpenses.length,
      expensesTotal: selectExpensesTotal(visibleExpenses)
    };
  })

  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount}</span> {expenseWord} totalling{' '}
          <span>{formattedExpensesTotal}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

export default ExpensesSummary;
