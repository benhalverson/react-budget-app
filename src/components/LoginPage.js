import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
  <div>
    <p>This is a demo app using firebase and google to login.</p>
    <p>Once signed in you can add edit and remove expenses.</p>
    <p>
      The datepicker is the AirBnB datepicker that is slightly modified to allow
      for selecting dates in the past.
    </p>
    <button onClick={startLogin}>Login</button>
  </div>
);

const mapDispatchToProps = dispatch => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(
  undefined,
  mapDispatchToProps
)(LoginPage);
