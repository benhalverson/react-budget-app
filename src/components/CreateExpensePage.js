import React, { Component} from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class CreateExpensePage extends Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

export default connect(undefined, { startAddExpense })(CreateExpensePage);
