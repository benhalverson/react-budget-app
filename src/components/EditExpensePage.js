import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensesPage extends Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, {
  startEditExpense, startRemoveExpense
})(EditExpensesPage);
