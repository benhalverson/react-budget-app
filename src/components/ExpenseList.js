import React from 'react';
import { useSelector } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = () => {
  const expenses = useSelector(({ expenses, filters }) =>
    selectExpenses(expenses, filters)
  );

  return (
    <div>
      {expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        expenses.map((expense) => (
          <ExpenseListItem key={expense.id} expense={expense} />
        ))
      )}
    </div>
  );
};

export default ExpenseList;
