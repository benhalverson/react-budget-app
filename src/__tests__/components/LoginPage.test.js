import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

test('should correctly render LoginPage', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});