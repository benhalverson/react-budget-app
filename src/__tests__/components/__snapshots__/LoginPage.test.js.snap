// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render LoginPage 1`] = `ShallowWrapper {}`;
